Fixed Referenced Style Broken Link

Shivam-1M opened the pull request:
👋 Hello there! Welcome. Please follow the steps below to tell us about your contribution.

1. Please complete a Checklist
2. Fill in all sections of the template
3. Click "Create pull request"

## Checklist

 <-- Please fill the boxes with [x] before submitting a pull request --> 

- [x] I have read the [How to Contribute](https://developers.cardano.org/docs/portal-contribute/).
- [x] I have run `yarn build` after adding my changes **without getting any errors**. 

## Updating documentation or Bugfix

<-- Help us understand your motivation by explaining why you decided to make this change. We must be able to understand the purpose of your change from this description.  Does this fix a bug? Does it close an issue? If so please mention it.  -->

Fixed the issue mentioned in #1229 (Reference Style Links Broken):
1. Works correctly for the referenced style syntax **"[.*?]"**.

> For example: [CIP Tag Registry Schema] mentioned in CIP-0114

2. Needed work on referenced style syntax **"[][.*?]"**

> For example: [registry][CIP Tag Registry] mentioned in CIP-0114

rphair commented:
Editing PR title a bit & marking this as a potential fix for  #1229.

rphair commented on .github/PULL_REQUEST_TEMPLATE/standard-change.md:
```diff
@@ -8,9 +8,9 @@
 
-<-- Help us understand your motivation by explaining why you decided to make this change. We must be able to understand the purpose of your change from this description.  Does this fix a bug? Does it close an issue? If so please mention it.  -->
\ No newline at end of file
+<-- Help us understand your motivation by explaining why you decided to make this change. We must be able to understand the purpose of your change from this description.  Does this fix a bug? Does it close an issue? If so please mention it.  -->
```
Looks like the last newline at the end of this file was either added or removed: difficult to fix from the GitHub editor so I think you will have to submit this change yourself from your local git.

In any case before we merge this PR there should be no changes to the file `.github/PULL_REQUEST_TEMPLATE/standard-change.md`.

Shivam-1M replied:
> Looks like the last newline at the end of this file was either added or removed: difficult to fix from the GitHub editor so I think you will have to submit this change yourself from your local git.
> 
> In any case before we merge this PR there should be no changes to the file `.github/PULL_REQUEST_TEMPLATE/standard-change.md`.

Apologies for that. I will fix that on my end.

Shivam-1M replied:
I pushed changes fixing the standard-change @rphair.

rphair replied:
thanks; looks good now.

rphair reviewed (commented):
Thanks so much @Shivam-1M for contributing this.  I've verified that it fixes the problem I reported with the Dev Portal build of CIP-0114 in https://github.com/cardano-foundation/developer-portal/issues/1229#issue-2197947613 - i.e. that the links to those JSON files resolve correctly from the footnote.

Before I recommend that we merge this, can you explain why we still get the warning message about those links being broken when building the Dev Portal locally?  (I think a complete fix would have to suppress the warning message in cases where the link is able to correctly resolve):
```
[WARNING] Docusaurus found broken links! ...
- On source page path = /docs/governance/cardano-improvement-proposals/CIP-0114:
   -> linking to ./registry.json (resolved as: /docs/governance/cardano-improvement-proposals/registry.json)
```

Shivam-1M commented:
> Thanks so much @Shivam-1M for contributing this. I've verified that it fixes the problem I reported with the Dev Portal build of CIP-0114 in [#1229 (comment)](https://github.com/cardano-foundation/developer-portal/issues/1229#issue-2197947613) - i.e. that the links to those JSON files resolve correctly from the footnote.
> 
> Before I recommend that we merge this, can you explain why we still get the warning message about those links being broken when building the Dev Portal locally? (I think a complete fix would have to suppress the warning message in cases where the link is able to correctly resolve):
> 
> ```
> [WARNING] Docusaurus found broken links! ...
> - On source page path = /docs/governance/cardano-improvement-proposals/CIP-0114:
>    -> linking to ./registry.json (resolved as: /docs/governance/cardano-improvement-proposals/registry.json)
> ```

It is pleasure working on it, no worries at all. Thank you for cheering up @rphair.

The warning message you mentioned persists **due to the use of various referenced style link formats in CIPs and CPS markdown**. In markdown, we can reference a link in the following ways:

1. Inline Links: Directly include the URL within the link text. 
`[cardano-foundation/CIPs](https://github.com/cardano-foundation/CIPs/tree/master/./CIP-0114/README.md))`
2. Reference-Style Links: Define the URL separately, referenced by a key.
`[CIP Tag Registry Schema]: ./registry.schema.json`
3. Mixed Reference Links: Use a combination where the link text refers to a reference key which is defined elsewhere.
`[registry][CIP Tag Registry]`

Our recent fix addressed issues with inline and reference-style links but didn't fully cover mixed reference links, where links like [registry][CIP Tag Registry] need to be resolved correctly. I'll work on extending the solution to handle these mixed references, ensuring they also resolve without warnings.

Once this is addressed, the warning message should no longer appear, and all links will be correctly resolved.

Shivam-1M commented:
> Also, @Shivam-1M: should there not be a corresponding change to `/scripts/cps.ts`... since footnoted links to asset files are just as likely for CPS documents as for CIPs?

Yes, you are absolutely correct. Changes to `/scripts/cps.ts` should also be incorporated just like CIPs.

Let me try and see how I can add a reusable function for both CPS and CIPs.

katomm reviewed (approved):
Thank you. Looks like a good improvement. Might makes sense to also check  if (referenceLinks.length > 0)

Shivam-1M commented:
**Mixed Reference Links Problem has also been solved**. `[registry][CIP Tag Registry]`
The new block of code ensures that mixed reference links are correctly resolved by combining the base URL with the relative path just as we did it with referenced-style links. @rphair 

Also, the `referenceLinks.length > 0` is correctly placed. Thanks for the suggestion @katomm.

Shivam-1M commented:
I noticed there are no mixed style links in any CPS markdown file, and there are no .json files either. 
Only one CPS ([CPS-0010](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md)) is using some .png files and the case is exclusively handled in the code in the file cps.ts:
```
const newContentWithCorrectImagePaths = newContentWithInfo.replace(
    /!\[.*?\]\(\.\/(.*?)\)/g,
    `![same text](../../../static/img/cps/${folderName}/$1)`
  );
```
The code in cps.ts does not handle markdown reference or inline links, likely because CPS documents do not include extra files besides .png files linked to the markdown files.

Do you want me to implement all CIP markdown functionalities to CPS markdown? Or is it better to move on as the situation demands?

rphair commented:
@Shivam-1M it's only a matter of time before a CPS is submitted that has mixed style links and/or JSON files.  The narrative style of the Cardano Problem Statement means these types of links would be less common, but they're not prohibited and these features would be useful in some cases that I am sure will be included in CPSs in the days ahead.

So, while we are fortunate enough to have your help and while this code is still fresh in your own mind, I would really appreciate if you would replicate _all_ the CIP functionality in the CPS processing as well.  Ideally these 2 types of documents should have exactly the same gamut of expression and interpretation (except for the YAML headers of course). 🙏  cc @katomm @fill-the-fill

Shivam-1M commented:
@rphair, thank you for your feedback and for highlighting the importance of consistency between CIP and CPS processing.

I am pleased to inform you that I have implemented all the functionalities from CIP processing into the CPS processing as well. This includes handling mixed style links, JSON files, and ensuring all link types are correctly processed.

I have thoroughly tested the updated cps.ts file with a CIP-0114 markdown link, and it worked perfectly. The CPS documents now support the same range of expressions and interpretations as the CIP documents, maintaining consistency across both types of documents. 

Please let me know if there are any other adjustments or improvements needed. 🙏

cc @katomm @fill-the-fill

rphair reviewed (commented):
thanks again & well done @Shivam-1M 🙏 I don't have the skill or experience to completely follow the code changes but I've tested the problematic CIPs (including CIP-0050 and CIP-0114 identfied as Dev Portal problems) and others in a local build and most of the broken links & all of the warnings are gone. 🎉

There is only one case I've found (CIP-0088) where valid links generating no warnings from Docusaurus are broken... when the link is to a directory, which Docusaurus is supposed to parse as a document link to that directory's `README.md`, as in the Rationale column of this table (link displays bare text `404: Not Found`):
- https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0088/#6-cip-specific-information
- https://github.com/cardano-foundation/CIPs/blob/c5f40f4c750f75fa89b9f1ee5c4e253aee32878d/CIP-0088/README.md?plain=1#L257

This issue was investigated with https://github.com/cardano-foundation/developer-portal/pull/1122 and was initially corrected upstream with https://github.com/cardano-foundation/CIPs/pull/773 - this also might not be a general problem & specifically related to the table context, while working normally elsewhere (?).

If you think this is unrelated to the work you are doing and/or this PR then @Shivam-1M please say so and I'll open a new issue to fix with a code change here, given that the upstream correction in https://github.com/cardano-foundation/CIPs/pull/773 could not solve the problem completely.  (cc @katomm @fill-the-fill)

katomm reviewed (approved):
LGTM, thank you so much.

Shivam-1M commented:
Thank you for the thorough testing and feedback, @rphair. I’m glad to hear that most of the broken links and warnings have been resolved.

**Regarding the issue with CIP-0088** and the links to directories, I can certainly investigate and work on resolving this. It seems like a specific edge case, especially given the context of the table, but I believe we can implement a solution to handle directory links correctly, ensuring they point to the README.md files as expected.

I'll start working on this issue and will update you once I have a solution. Thank you for bringing this to my attention.

cc @katomm @fill-the-fill

Shivam-1M commented:
Issue with CIP-0088 and for all future nested CIP directory links have been resolved. Please take a look and let me know if it helps or need any updates.

Thanks.

cc @rphair @katomm @fill-the-fill

rphair reviewed (approved):
nice work @Shivam-1M ... asset subdirectories for CIP-0088 now link to directory pages on GitHub 🤓
